README
			   ABOUT PY-NOTIFY

Py-notify is a Python package providing tools for implementing
Observer programming pattern. These tools include signals, conditions
and variables.

Signals are lists of handlers that are called when signal is
emitted. Conditions are basically boolean variables coupled with a
signal that is emitted when condition state changes. They can be
combined using standard logical operators (not, and, etc.) into
compound conditions. Variables, unlike conditions, can hold any Python
object, not just booleans, but they cannot be combined.

All three main concepts support deriving by users. Conditions and
variables even have derive_type() method that can be used to subclass
them in just a few lines of code, for instance to create a variable
that restricts its value's type.

Latest Py-notify versions can be found at the homepage:

		    http://home.gna.org/py-notify/



			RUNNING STANDARD TESTS

Before installing Py-notify, please run the standard test cases.  It
is very simple, just execute

	./run-tests.py

in this directory.  Tests take just a few seconds to complete.  If any
of them fails in a released version, it is a bug, so please report it.
However, this doesn't apply to Py-notify you get from Subversion
repository, those can occasionally contain failing tests.



			     INSTALLATION

The usual `./setup.py build; ./setup.py install' or just `./setup.py
install' should build and install Py-notify.  See file INSTALL for
details.



			    REPORTING BUGS

Please use bug tracker at the Gna! project page for bug submissions.
Point your browser to

		https://gna.org/bugs/?group=py-notify

Before submitting a bug, please update to the latest SVN version of
Py-notify and verify that the bug is still present.  Don't forget to
check if the bug has been submitted already.

Please include the following information in your bug report:

* Bug description and the way to reproduce the bug.

* Exception stack trace, if any.

* System information, most importantly Python version.

* Test case if possible.  If a standard test fails (one of those in
  the `test' directory), its name in stack trace will suffice.

* Everything else you find relevant.



		       COPYRIGHTS AND LICENSES

All files are copyrighted by Paul Pogonyshev except as noted below.
See each specific file for details.

All source files are licensed under the GNU Lesser General Public
License version 2.1 or (at your option) any later version.  GNU LGPL
version 2.1 can be found in the file COPYING.

* File `setup.py' is explicitly put in Public Domain.

* File `benchmark/configobj.py' is copyright (C) 2005-2006 Michael
  Foord, Nicola Larosa and released under the terms of the BSD
  License.


NEWS
--- New in Py-notify 0.3.2  (??? -- in development) ------------------

* Part of functionality of `notify.gc' module is now implemented in
  Python, not C.


--- New in Py-notify 0.3.1  (28 September 2008) ----------------------

* All methods that accepted arbitrary argument numbers now also accept
  arbitrary keyword arguments.

* Updated to compile and work with Python 3.0rc1.

* A few minor fixes for potential bugs and optimizations in
  `notify.gc' module and in documentation.

* 6 more unit test (164/186 in total).


######################################################################


--- New in Py-notify 0.2.1  (21 September 2008) ----------------------

* Updated to compile and work with Python 3.0rc1.

* A few minor fixes for potential bugs and optimizations in
  `notify.gc' module and in documentation.


### New in Py-notify 0.2.0  (6 May 2008) #############################

* First stable version.  158/180 unit tests.  All pass with Python
  2.3, 2.5, SVN trunk (heading to 2.6) and SVN py3k branch on Debian
  GNU/Linux.

* One more improvement in test framework.  Tiny improvement in
  `run-tests.py' for Py3k.

######################################################################


--- New in Py-notify 0.1.15  (3 March 2008) --------------------------

* Important bug-fix in `CleanSignal'.

* `AbstractGCProtector.default' cannot be changed anymore if the
  current default protector is used.  In practice, it can be changed
  only on program startup.

* Changed internals of `AbstractGCProtector.default' (in backward
  compatible fashion) to avoid screwing class attribute cache in
  upcoming Python 2.6 (see http://bugs.python.org/issue1878).

* Documentation improvements.

* 1 more unit test (158/180 in total).  Much improved test framework
  catches more bugs for already written tests.


--- New in Py-notify 0.1.14  (28 December 2007) ----------------------

* Update to work with SVN Py3k again.

* Fix build on MinGW at least in some cases (#10560).

* A few documentation and infrastructure fixes and changes.


--- New in Py-notify 0.1.13  (17 November 2007) ----------------------

* Changed to work under SVN builds of Python 3000 too.  It still
  requires only 2.3 or later.  Note that one testcase fails under
  Py3k, but that is a Py3k bug: http://bugs.python.org/issue1393.
  [P.S. Py3k was fixed later.]

* A few minor bugfixes.

* 2 new unit tests (157/179 in total).


--- New in Py-notify 0.1.12  (25 August 2007) ------------------------

* New method AbstractVariable.transform().

* Some bugfixes and code improvements.

* 8 new unit tests (155/177 in total).


--- New in Py-notify 0.1.11  (22 July 2007) --------------------------

* Major speed optimization in signal emission.  Now all benchmarks run
  about 30% faster.

* Many small improvements and fixes in derive_type() machinery.

* Minor internal code improvements.

* 18 more unit tests (147/169 in total).


--- New in Py-notify 0.1.10  (1 July 2007) ---------------------------

* Possibility to freeze changes in conditions and variabes.

* For Python 2.5+ only: 8 context manager methods for signals,
  conditions and variables.

* Expanded tutorial.

* Other small improvements and fixes.

* 31 more unit tests (151 in total).  Now not all tests run under
  Python 2.3 or 2.4; for them there are only 129 tests.


--- New in Py-notify 0.1.9  (22 June 2007) ---------------------------

* Fixed and improved `setup.py' script.

* New RaisingGCProtector class.

* 6 new unit tests (120 in total).


--- New in Py-notify 0.1.8  (13 June 2007) ---------------------------

* Improved and fixed derive_type() class method in several
  AbstractValueObject-derived classes.

* Mediators and bindings got __hash__() method.

* Informative string representation added to bindings and
  DummyReference class, improved in signals.

* Other small improvements and fixes.

* 8 new unit tests (114 in total) and several improved.


--- New in Py-notify 0.1.7  (10 April 2007) --------------------------

* Improved documentation.  Some documentation files that didn't
  previously make it to source distribution now do.

* Many minor improvements and fixes.

* New `examples' subdirectory with one but non-trivial example.

* 1 removed and 9 new unit tests (106 in total).

* 1 new benchmark test (5 in total).


--- New in Py-notify 0.1.6  (4 April 2007) ---------------------------

* Adopted "Don't repeat yourself" principle, therefore most functions
  that duplicate properties are renamed and made private.  This breaks
  backward compatibility, but better now than never, since user base
  is nonexistent.

* Some bug-fixes and proper Pythonic __eq__, __ne__ and numeric
  methods.

* New DebugGCProtector class.

* Improved or added documentation for many methods.

* Infrastructure script improvements and fixes.

* 6 more unit tests (98 in total).


--- New in Py-notify 0.1.5  (3 March 2007) ---------------------------

* Added a tutorial.

* Improved or added documentation for many methods.

* New functions to cancel effects of synchronize() for condiions and
  variables.

* Several more properties in different classes.  In general, use
  properties instead of methods more.

* Some bug-fixes.

* 5 more unit tests (92 in total).


--- New in Py-notify 0.1.4  (16 February 2007) -----------------------

* New fast and generic GC-protector module (notify.gc) replaces
  inefficient mark_object_as_used() and mark_object_as_unused()
  functions.

* Speed optimizations in bindings.

* 2 more unit tests (87 in total).

* 2 more benchmark tests (4 in total).


--- New in Py-notify 0.1.3  (3 February 2007) ------------------------

* New class `WatcherVariable', just like `WatcherCondition'.

* Many bug-fixes in `Signal' class and in standard signal value
  accumulators.

* Speed optimizations in bindings and signals.

* 18 more unit tests (85 in total).



--- New in Py-notify 0.1.2  (28 January 2007) ------------------------


notify/utils.py
# -*- coding: utf-8 -*-

#--------------------------------------------------------------------#
# This file is part of Py-notify.                                    #
#                                                                    #
# Copyright (C) 2006, 2007, 2008 Paul Pogonyshev.                    #
#                                                                    #
# This library is free software; you can redistribute it and/or      #
# modify it under the terms of the GNU Lesser General Public License #
# as published by the Free Software Foundation; either version 2.1   #
# of the License, or (at your option) any later version.             #
#                                                                    #
# This library is distributed in the hope that it will be useful,    #
# but WITHOUT ANY WARRANTY; without even the implied warranty of     #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU  #
# Lesser General Public License for more details.                    #
#                                                                    #
# You should have received a copy of the GNU Lesser General Public   #
# License along with this library; if not, write to the Free         #
# Software Foundation, Inc., 51 Franklin Street, Fifth Floor,        #
# Boston, MA 02110-1301 USA                                          #
#--------------------------------------------------------------------#


"""
A collection of utilities that can also be used from outside, if wanted.  Functions and
classes here can be assumed public and won’t disappear in future Py-notify versions.

@var is_callable:
Determine if C{object} is callable.  E.g. if it is a function, method, class, instance of
a class with C{__call__}, etc.  This is the same as built-in function C{callable} does.
C{is_callable} is provided since C{callable} is going to disappear in Python 3000 and may
issue warnings in 2.6.

@var as_string:
Convert any attribute to its name as string.  Main use of this utility object is to
perform Python ‘private’ identifier mangling.  E.g. you can write::

    class MyClass (object):
        __slots__ = ('__x')
        def get_x (self):
            if hasattr (self, as_string.__x):
                return self.__x

Advantage is that you don’t have to do mangling ‘by hands’ and hence there is less chance
for a typing error.  Furthermore, this code does not require changes if you change
C{MyClass} name to anything else, whereas custom mangling does.

However, usefulness of ‘as_string’ is still doubtful.  When I wrote it, I didn’t know one
could just write ``__slots__ = ('__x')``, I thought it needed to be
``__slots__ = ('_MyClass__x')``.  Imagine...
"""

__docformat__ = 'epytext en'
__all__       = ('is_callable', 'is_valid_identifier', 'mangle_identifier',
                 'as_string',
                 'raise_not_implemented_exception',
                 'execute',
                 'frozendict', 'DummyReference', 'ClassTypes', 'StringType')


import re
import sys
import types
from keyword import iskeyword



if sys.version_info[:3] < (2, 6, 0):
    is_callable = callable

else:
    def is_callable (object):
        return hasattr (object, '__call__')



def is_valid_identifier (identifier):
    """
    Determine if C{identifier} is a valid Python identifier.  This function never raises
    any exceptions.  If C{identifier} is not a string, it simply returns C{False}.

    @param identifier: identifier to determin if it is valid
    @type  identifier: C{basestring}

    @rtype:            C{bool}
    """

    return (isinstance (identifier, StringType)
            and re.match ('^[_a-zA-Z][_a-zA-Z0-9]*$', identifier) is not None
            and not iskeyword (identifier))


def mangle_identifier (class_name, identifier):
    """
    Mangle C{identifier} as how would be done if it appeared in a class with
    C{class_name}.  This function allows to mimic standard Python mangling of
    pseudo-private attributes, i.e. those which names start with two underscores and don’t
    end in two.  If C{identifier} is not considered a private name, it is returned
    unchanged.

    @param  class_name: name of Python class.
    @type   class_name: C{basestring}

    @param  identifier: name of an attribute of that class.
    @type   identifier: C{basestring}

    @rtype: C{str}

    @raises ValueError: if either C{class_name} or C{identifier} is not valid from
                        Python’s point of view.
    """

    if not (is_valid_identifier (class_name) and is_valid_identifier (identifier)):
        raise ValueError ("'class_name' and 'identifier' must be valid Python identifiers")

    if (identifier.startswith ('__')
        and not identifier.endswith ('__')
        and class_name != '_' * len (class_name)):
        return '_%s%s' % (class_name.lstrip ('_'), identifier)
    else:
        return identifier



class _AsString (object):

    """
    Internal helper class for C{L{as_string}}.  Don’t use directly.
    """

    __slots__ = ()

    def __getattribute__(self, name):
        return name

    def __setattr__(self, name, value):
        raise TypeError ("'as_string' attributes cannot be set")

    def __delattr__(self, name):
        raise TypeError ("'as_string' attributes cannot be deleted")

    def __repr__(self):
        return 'notify.utils.as_string'


as_string = _AsString ()



def raise_not_implemented_exception (object = None, function_name = None):
    """
    Raise C{NotImplementedError} for a method invoked with C{object} as C{self}.  The
    function determines object class and method declaration class(es) itself and that’s
    the whole point of it.

    It should be called like this:
        >>> raise_not_implemented_exception (self)

    And output might look like this::
       File ".../foo.py", line # in ?
         Foo ().bar ()
       File ".../foo.py", line #, in bar
         raise_not_implemented_exception (self)
       File ".../notify/utils.py", line #, in raise_not_implemented_exception
         raise exception
     NotImplementedError: bar() not implemented in class Foo (declared in AbstractFoo)

    Optionally, C{function_name} can be specified.  This argument mainly exists for C
    extension, since function name cannot be detected automatically in this case.  In
    Python code you should just leave this argument out.

    @param  object:              the object for which a non-implemented method is called.
    @type   object:              C{object}

    @param  function_name:       name of the unimplemented function or method (inferred
                                 automatically for non-extension functions).
    @type   function_name:       C{basestring} or C{None}

    @raises NotImplementedError: always.
    """

    if function_name is None:
        try:
            raise Exception
        except Exception:
            try:
                traceback     = sys.exc_info () [2]
                function_name = traceback.tb_frame.f_back.f_code.co_name
            except Exception:
                # We can do nothing, ignore.
                pass

    if function_name is not None:
        function_description = '%s()' % function_name
    else:
        function_description = 'UNKNOWN FUNCTION'

    try:
        class_description = ' in class %s' % object.__class__.__name__

        if function_name is not None:
            declaration_classes = _find_declaration_classes (object.__class__, function_name)

            if len (declaration_classes) == 1:
                if declaration_classes[0] is not object.__class__:
                    class_description += ' (declared in %s)' % declaration_classes[0].__name__

            elif len (declaration_classes) > 1:
                class_description += (' (declared in %s)'
                                      % ', '.join ([_class.__name__
                                                    for _class in declaration_classes]))

    except Exception:
        class_description = ''

    exception = NotImplementedError ('%s not implemented%s'
                                     % (function_description, class_description))
    raise exception


def _find_declaration_classes (_class, function_name):
    declaring_bases = [base for base in _class.__bases__ if hasattr (base, function_name)]

    if declaring_bases:
        return reduce (lambda list1, list2: list1 + list2,
                       [_find_declaration_classes (base, function_name)
                        for base in declaring_bases],
                       [])
    else:
        return [_class]


if sys.version_info[0] >= 3:
    execute = eval ('exec')
else:
    from notify._2_x import execute



class frozendict (dict):

    __slots__ = ('__hash')

    def __init__(self, *arguments, **keywords):
        super (frozendict, self).__init__(*arguments, **keywords)
        self.__hash = None


    def clear (self):
        raise TypeError ("'%s' object doesn't support clearing" % type (self).__name__)

    def pop (self, key, default = None):
        raise TypeError ("'%s' object doesn't support popping" % type (self).__name__)

    def popitem (self):
        raise TypeError ("'%s' object doesn't support popping" % type (self).__name__)

    def setdefault (self, key, default = None):
        raise TypeError ("'%s' object doesn't support setdefault operation" % type (self).__name__)

    def update (self, dict):
        raise TypeError ("'%s' object doesn't support updating" % type (self).__name__)


    def __setitem__(self, key, value):
        raise TypeError ("'%s' object doesn't support item setting" % type (self).__name__)

    def __delitem__(self, key):
        raise TypeError ("'%s' object doesn't support item deletion" % type (self).__name__)


    def __hash__(self):
        _hash = self.__hash

        if _hash is None:
            _hash = 0x1337

            if hasattr (dict, 'iteritems'):
                for key, value in self.iteritems ():
                    _hash ^= hash (key) ^ hash (value)
            else:
                for key, value in self.items ():
                    _hash ^= hash (key) ^ hash (value)

            self.__hash = _hash

        return _hash


    def __repr__(self):
        return '%s (%s)' % (type (self).__name__, super (frozendict, self).__repr__())



frozendict.EMPTY = frozendict ({ })
# Force hash to be precomputed.
hash (frozendict.EMPTY)



class DummyReference (object):

    """
    Simple class that is interface-compatible with C{weakref.ReferenceType}.  In other
    words, its constructor accepts only one parameter and this value is later returned
    from C{L{__call__}} method.  Unlike weak references, instances of this class don’t do
    anything special.  They are only needed to avoid special cases for non-references,
    since you can treat instances of C{weakref.ReferenceType} and this class in the same
    way.
    """

    __slots__ = ('__object')


    def __init__(self, object):
        """
        Create a new dummy reference that will return C{object} when called.

        @param object: the object that will be returned by this reference.
        @type  object: C{object}
        """

        self.__object = object

    def __call__(self):
        """
        Return the C{object} specified at construction time.

        @rtype: C{object}
        """

        return self.__object


    def __repr__(self):
        return ('<%s.%s at 0x%x; to %r>'
                % (self.__module__, self.__class__.__name__, id (self), self.__object))

    def __str__(self):
        return '<%s at 0x%x; to %s>' % (self.__class__.__name__, id (self), self.__object)


if sys.version_info[0] >= 3:
    ClassTypes = (type,)
    StringType = str
else:
    ClassTypes = (type, types.ClassType)
    StringType = basestring



# Local variables:
# mode: python
# python-indent: 4
# indent-tabs-mode: nil
# fill-column: 90
# End:
